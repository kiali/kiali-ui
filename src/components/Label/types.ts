export interface Labels {
  name: string;
  value: string;
}
