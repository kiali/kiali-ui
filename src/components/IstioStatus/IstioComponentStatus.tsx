import * as React from 'react';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { IconType } from '@patternfly/react-icons/dist/js/createIcon';
import { PFAlertColor } from '../Pf/PfColors';
import { CheckCircleIcon, ExclamationCircleIcon, ExclamationTriangleIcon } from '@patternfly/react-icons';
import { Split, SplitItem } from '@patternfly/react-core';

type Props = {
  componentStatus: ComponentStatus;
};

export type ComponentIcon = {
  color: string;
  icon: IconType;
};

const ErrorCoreComponent: ComponentIcon = {
  color: PFAlertColor.Danger,
  icon: ExclamationCircleIcon
};

const ErrorAddonComponent: ComponentIcon = {
  color: PFAlertColor.Warning,
  icon: ExclamationTriangleIcon
};

const SuccessComponent: ComponentIcon = {
  color: PFAlertColor.Success,
  icon: CheckCircleIcon
};

// Mapping Valid-Core to Icon representation.
const validToIcon: { [valid: string]: ComponentIcon } = {
  'false-false': ErrorAddonComponent,
  'false-true': ErrorCoreComponent,
  'true-false': SuccessComponent,
  'true-true': SuccessComponent
};

const statusMsg = {
  [Status.NotFound]: 'Not found',
  [Status.Unhealthy]: 'Not healthy',
  [Status.Unreachable]: 'Unreachable'
};

class IstioComponentStatus extends React.Component<Props> {
  renderIcon = (status: Status, isCore: boolean) => {
    const compIcon = validToIcon[`${status === Status.Healthy}-${isCore}`];
    const IconComponent = compIcon.icon;
    return <IconComponent style={{ color: compIcon.color, marginTop: 5 }} />;
  };

  renderCells = () => {
    const comp = this.props.componentStatus;

    return [
      <Split key={'cell-status-icon-' + comp.name} gutter={'md'}>
        <SplitItem>{this.renderIcon(this.props.componentStatus.status, this.props.componentStatus.is_core)}</SplitItem>
        <SplitItem isFilled={true}>{comp.name}</SplitItem>
        <SplitItem>{statusMsg[comp.status]}</SplitItem>
      </Split>
    ];
  };

  render() {
    return this.renderCells();
  }
}

export default IstioComponentStatus;
