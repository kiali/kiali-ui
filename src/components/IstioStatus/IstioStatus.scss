#istio-status {
  &.pf-c-data-list {
    color: var(--pf-c-tooltip__content--Color);
    --pf-c-data-list--BorderTopWidth: 0;
    margin: 0;
    padding: 0;
  }

  .pf-c-data-list__item {
    --pf-c-data-list--BackgroundColor: var(--pf-c-tooltip__content--BackgroundColor);
    --pf-c-data-list__item-item--BorderTopWidth: 0;
  }

  .pf-c-data-list__item-row {
    --pf-c-data-list__item-row--PaddingRight: 0;
    --pf-c-data-list__item-row--PaddingLeft: 0;
  }

  .pf-c-data-list__cell {
    --pf-c-data-list__cell--PaddingTop: 0;
    --pf-c-data-list__cell-cell--PaddingTop: 0;
  }

  .pf-c-data-list__item-content {
    --pf-c-data-list__item-content--PaddingBottom: 0;
  }

  &.pf-c-list {
    --pf-c-content--ul--PaddingLeft: 0;
    --pf-c-content--ul--MarginTop: 0;
    --pf-c-content--ul--MarginLeft: 0;

    .pf-l-split {
      color: var(--pf-c-tooltip__content--Color);

      .pf-l-split__item {
        text-align: left;
      }
    }
  }
}
