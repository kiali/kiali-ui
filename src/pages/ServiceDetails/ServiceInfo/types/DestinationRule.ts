import { DestinationRuleSpec, K8sMetadata } from '../../../../types/IstioObjects';
import Subset from './Subset';

export default class DestinationRule {
  metadata: K8sMetadata;
  spec: DestinationRuleSpec;
  unformattedField: string;

  constructor(metadata: K8sMetadata, spec: DestinationRuleSpec) {
    this.metadata = metadata;
    this.spec = spec;
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.spec !== null;
  }

  hasValidSubsets() {
    if (!this.spec.subsets) {
      return true;
    }

    let valid = this.spec.subsets instanceof Array;
    valid =
      valid &&
      this.spec.subsets.every((subset, i, ary) =>
        new Subset(subset.name, subset.labels, subset.trafficPolicy).isValid()
      );

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
