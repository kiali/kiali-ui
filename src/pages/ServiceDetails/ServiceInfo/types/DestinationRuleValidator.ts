import { DestinationRule } from '../../../../types/IstioObjects';
import SubsetValidator from './SubsetValidator';

export default class DestinationRuleValidator {
  destinationRule: DestinationRule;
  unformattedField: string;

  constructor(destinationRule: DestinationRule) {
    this.destinationRule = destinationRule;
    this.unformattedField = 'none';
  }

  isValid() {
    return this.hasValidName() && this.hasSpecData() && this.hasValidSubsets() && this.hasValidHost();
  }

  hasValidName() {
    if (typeof this.destinationRule.metadata.name !== 'string') {
      this.unformattedField = 'Name';
      return false;
    }

    return true;
  }

  hasSpecData() {
    return this.destinationRule.spec !== null;
  }

  hasValidSubsets() {
    if (!this.destinationRule.spec.subsets) {
      return true;
    }

    let valid = this.destinationRule.spec.subsets instanceof Array;
    valid =
      valid && this.destinationRule.spec.subsets.every((subset, _i, _ary) => new SubsetValidator(subset).isValid());

    if (!valid) {
      this.unformattedField = 'Subsets';
    }

    return valid;
  }

  hasValidHost() {
    if (typeof this.destinationRule.spec.host !== 'string') {
      this.unformattedField = 'Host';
      return false;
    }

    return true;
  }

  formatValidation() {
    if (!this.isValid()) {
      return {
        message: 'This destination rule has format problems in field ' + this.unformattedField,
        severity: 'error',
        path: ''
      };
    }

    return null;
  }
}
