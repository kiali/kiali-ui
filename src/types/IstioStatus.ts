export enum Status {
  Healthy = 'Healthy',
  Unhealthy = 'Unhealthy',
  Unreachable = 'Unreachable',
  NotFound = 'NotFound'
}

export interface ComponentStatus {
  name: string;
  status: Status;
  is_core: boolean;
}
