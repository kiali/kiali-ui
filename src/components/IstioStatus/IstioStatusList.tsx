import * as React from 'react';
import { List, Text, TextContent, TextVariants } from '@patternfly/react-core';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import IstioComponentStatus from './IstioComponentStatus';
import { PfColors } from '../Pf/PfColors';

type Props = {
  status: ComponentStatus[];
};

class IstioStatusList extends React.Component<Props> {
  nonhealthyComponents = () => {
    return this.props.status.filter((c: ComponentStatus) => c.status !== Status.Healthy);
  };

  coreComponentsStatus = () => {
    return this.nonhealthyComponents().filter((s: ComponentStatus) => s.is_core);
  };

  addonComponentsStatus = () => {
    return this.nonhealthyComponents().filter((s: ComponentStatus) => !s.is_core);
  };

  renderComponentList = () => {
    const groups = {
      core: this.coreComponentsStatus,
      addon: this.addonComponentsStatus
    };

    return ['core', 'addon'].map((group: string) => {
      return (
        <>
          {groups[group]().map(status => {
            return <IstioComponentStatus componentStatus={status} />;
          })}
        </>
      );
    });
  };

  render() {
    return (
      <TextContent style={{ color: PfColors.White }}>
        <Text component={TextVariants.h4}>Istio Components Status</Text>
        <List id="istio-status" aria-label="Istio Component List">
          {this.renderComponentList()}
        </List>
      </TextContent>
    );
  }
}

export default IstioStatusList;
