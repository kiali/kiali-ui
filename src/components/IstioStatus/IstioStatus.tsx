import * as React from 'react';
import * as API from '../../services/Api';
import * as AlertUtils from '../../utils/AlertUtils';
import { TimeInMilliseconds } from '../../types/Common';
import { ComponentStatus, Status } from '../../types/IstioStatus';
import { MessageType } from '../../types/MessageCenter';
import Namespace from '../../types/Namespace';
import { KialiAppState } from '../../store/Store';
import { istioStatusSelector, lastRefreshAtSelector, namespaceItemsSelector } from '../../store/Selectors';
import { KialiDispatch } from '../../types/Redux';
import { bindActionCreators } from 'redux';
import { IstioStatusActions } from '../../actions/IstioStatusActions';
import { connect } from 'react-redux';
import { Tooltip, TooltipPosition } from '@patternfly/react-core';
import IstioStatusList from './IstioStatusList';
import { PFAlertColor } from '../Pf/PfColors';
import './IstioStatus.css';
import { ResourcesFullIcon } from '@patternfly/react-icons';

type ReduxProps = {
  lastRefreshAt: TimeInMilliseconds;
  setIstioStatus: (istioStatus: ComponentStatus[]) => void;
  namespaces: Namespace[] | undefined;
  status: ComponentStatus[];
};

type Props = ReduxProps & {};

const ValidToColor = {
  'false-false': PFAlertColor.Danger,
  'false-true': PFAlertColor.Danger,
  'true-false': PFAlertColor.Warning,
  'true-true': PFAlertColor.Success
};

export class IstioStatus extends React.Component<Props> {
  componentDidMount() {
    this.fetchStatus();
  }

  componentDidUpdate(prevProps: Readonly<Props>): void {
    if (this.props.lastRefreshAt !== prevProps.lastRefreshAt) {
      this.fetchStatus();
    }
  }

  fetchStatus = () => {
    API.getIstioStatus()
      .then(response => {
        return this.props.setIstioStatus(response.data);
      })
      .catch(error => {
        // User without namespaces can't have access to mTLS information. Reduce severity to info.
        const informative = this.props.namespaces && this.props.namespaces.length < 1;
        if (informative) {
          AlertUtils.addError('Istio deployment status disabled.', error, 'default', MessageType.INFO);
        } else {
          AlertUtils.addError('Error fetching Istio deployment status.', error, 'default', MessageType.ERROR);
        }
      });
  };

  tooltipContent = () => {
    return <IstioStatusList status={this.props.status} />;
  };

  tooltipColor = () => {
    let coreHealthy: boolean = true;
    let addonHealthy: boolean = true;

    Object.keys(this.props.status || {}).forEach((compKey: string) => {
      const { status, is_core } = this.props.status[compKey];
      const isHealthy: boolean = status === Status.Healthy;

      if (is_core) {
        coreHealthy = coreHealthy && isHealthy;
      } else {
        addonHealthy = addonHealthy && isHealthy;
      }
    });

    return ValidToColor[`${coreHealthy}-${addonHealthy}`];
  };

  healthyComponents = () => {
    return this.props.status.reduce((healthy: boolean, compStatus: ComponentStatus) => {
      return healthy && compStatus.status === Status.Healthy;
    }, true);
  };

  render() {
    if (!this.healthyComponents()) {
      return (
        <Tooltip position={TooltipPosition.left} enableFlip={true} content={this.tooltipContent()} maxWidth={'25rem'}>
          <ResourcesFullIcon color={this.tooltipColor()} style={{ verticalAlign: '-0.3em', marginRight: 8 }} />
        </Tooltip>
      );
    }

    return null;
  }
}

const mapStateToProps = (state: KialiAppState) => ({
  status: istioStatusSelector(state),
  lastRefreshAt: lastRefreshAtSelector(state),
  namespaces: namespaceItemsSelector(state)
});

const mapDispatchToProps = (dispatch: KialiDispatch) => ({
  setIstioStatus: bindActionCreators(IstioStatusActions.setinfo, dispatch)
});

const IstioStatusConnected = connect(
  mapStateToProps,
  mapDispatchToProps
)(IstioStatus);

export default IstioStatusConnected;
