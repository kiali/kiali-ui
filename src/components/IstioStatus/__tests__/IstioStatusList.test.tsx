import * as React from 'react';
import { shallow } from 'enzyme';
import { ComponentStatus, Status } from '../../../types/IstioStatus';
import IstioStatusList from '../IstioStatusList';
import { shallowToJson } from 'enzyme-to-json';

it('lists all the components grouped', () => {
  const components: ComponentStatus[] = [
    {
      name: 'grafana',
      status: Status.NotFound,
      is_core: false
    },
    {
      name: 'prometheus',
      status: Status.Unhealthy,
      is_core: false
    },
    {
      name: 'istiod',
      status: Status.NotFound,
      is_core: true
    },
    {
      name: 'istio-egressgateway',
      status: Status.Unhealthy,
      is_core: true
    }
  ];

  const wrapper = shallow(<IstioStatusList status={components} />);

  expect(shallowToJson(wrapper)).toBeDefined();
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});
