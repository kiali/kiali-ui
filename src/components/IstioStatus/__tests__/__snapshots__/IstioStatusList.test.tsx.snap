// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lists all the components grouped 1`] = `
<TextContent
  style={
    Object {
      "color": "#fff",
    }
  }
>
  <Text
    component="h4"
  >
    Istio Components Status
  </Text>
  <List
    aria-label="Istio Component List"
    id="istio-status"
  >
    <IstioComponentStatus
      componentStatus={
        Object {
          "is_core": true,
          "name": "istiod",
          "status": "NotFound",
        }
      }
    />
    <IstioComponentStatus
      componentStatus={
        Object {
          "is_core": true,
          "name": "istio-egressgateway",
          "status": "Unhealthy",
        }
      }
    />
    <IstioComponentStatus
      componentStatus={
        Object {
          "is_core": false,
          "name": "grafana",
          "status": "NotFound",
        }
      }
    />
    <IstioComponentStatus
      componentStatus={
        Object {
          "is_core": false,
          "name": "prometheus",
          "status": "Unhealthy",
        }
      }
    />
  </List>
</TextContent>
`;
